redwood-protocol-host/src/commonMain/kotlin/app/cash/redwood/protocol/widget/ProtocolBridge.kt

```diff
@@ -137,6 +169,225 @@ public class ProtocolBridge<W : Any>(
   private fun node(id: Id): ProtocolNode<W> {
     return checkNotNull(nodes[id]) { "Unknown widget ID ${id.value}" }
   }
+
+  /**
+   * Implements widget reuse (view recycling).
+   *
+   * When a widget is eligible from reuse:
+   *
+   *  * It is removed from [pool].
+   *  * It is added to [nodes], alongside its descendant nodes.
+   *
+   * Returns the updated set of changes that omits any changes that were implemented with reuse.
+   */
+  private fun applyReuse(changes: List<Change>): List<Change> {
+    val idToSubtree = mutableMapOf<Id, ReuseSubtree<W>>()
+
+    // Find nodes that have Modifier.reuse
+    for ((index, change) in changes.withIndex()) {
+      if (change !is ModifierChange) continue
+
+      val reuseModifierIndex = change.elements.indexOfFirst { it.tag.value == REUSE_MODIFIER_TAG }
+      if (reuseModifierIndex == -1) continue
+
+      val candidateIndex = pool.indexOfFirst { it.reuse }
+      if (candidateIndex == -1) continue
+
+      if (changes.subList(0, index).none { it is Create && it.id == change.id }) continue
```

Comment:
I think this is accidentally quadratic? At least in the case of enough reuse. Since the protocol guarantees a create immediately precedes the property+modifier assignment, we can walk backwards from `index` until we see either a `Create` (and `continue`) or `change.id != it.id` (and stop looking). This should minimize the look-behind to a few changes at most.

Reply:
Great call. I think there’s an entire follow-up PR where I take this kinda-brute-force approach and tune it for some big efficiency gains. If you don’t mind I have a few such changes in mind and I’d like to do them in a follow-up.

(This PR: it works. Next PR: it still works, and it’s a bit faster.)